### PR TITLE
[WM-1646] POST /method endpoint

### PR DESCRIPTION
**Pull request by salonishah11**
Request to endpoint:
```
{
  "method_name": "fetch_sra_to_bam method",
  "method_description": "fetch_sra_to_bam method description",
  "method_source": "GitHub",
  "method_version": "master",
  "method_url": "https://raw.githubusercontent.com/broadinstitute/viral-pipelines/master/pipes/WDL/workflows/fetch_sra_to_bam.wdl"
}
```
Response from endpoint upon successful creation:
```
{
  "method_id": "19def919-5e62-4064-92ac-79392a7d887f",
  "run_set_id": "9af55416-d51b-4510-8d75-09ae91c21598"
}
```
Database entries:
```
cbas_db=# select * from method where method_id='19def919-5e62-4064-92ac-79392a7d887f';
-[ RECORD 1 ]---+-------------------------------------
method_id       | 19def919-5e62-4064-92ac-79392a7d887f
name            | fetch_sra_to_bam method
description     | fetch_sra_to_bam method description
created         | 2023-02-07 19:34:52.613836-05
method_source   | GitHub
last_run_set_id | 9af55416-d51b-4510-8d75-09ae91c21598

cbas_db=# 
cbas_db=# select * from method_version where method_id='19def919-5e62-4064-92ac-79392a7d887f';
-[ RECORD 1 ]------------------+-----------------------------------------------------------------------------------------------------------------
method_version_id              | c369d581-360e-437d-b51b-819d63b02356
method_id                      | 19def919-5e62-4064-92ac-79392a7d887f
method_version_name            | master
method_version_description     | fetch_sra_to_bam method description
method_version_created         | 2023-02-07 19:34:52.613902-05
method_version_last_run_set_id | 9af55416-d51b-4510-8d75-09ae91c21598
method_version_url             | https://raw.githubusercontent.com/broadinstitute/viral-pipelines/master/pipes/WDL/workflows/fetch_sra_to_bam.wdl

cbas_db=# 
cbas_db=# select * from run_set where run_set_id='9af55416-d51b-4510-8d75-09ae91c21598';
-[ RECORD 1 ]-----------+--------------------------------------------------------------------
run_set_id              | 9af55416-d51b-4510-8d75-09ae91c21598
is_template             | t
run_set_name            | fetch_sra_to_bam method/master workflow
run_set_description     | Template Run Set for Method fetch_sra_to_bam method/master workflow
status                  | COMPLETE
submission_timestamp    | 2023-02-07 19:34:52.614307-05
last_modified_timestamp | 2023-02-07 19:34:52.614311-05
last_polled_timestamp   | 2023-02-07 19:34:52.614313-05
run_count               | 0
error_count             | 0
input_definition        | {}
output_definition       | {}
record_type             | 
method_version_id       | c369d581-360e-437d-b51b-819d63b02356
```

Closes https://broadworkbench.atlassian.net/browse/WM-1646

### CODE REVIEW

File: service/src/main/java/bio/terra/cbas/controllers/MethodsApiController.java [hunk 1]
```diff
@@ -1,31 +1,163 @@
...
 import org.springframework.stereotype.Controller;
 
 @Controller
 public class MethodsApiController implements MethodsApi {
+  private final CromwellService cromwellService;
   private final MethodDao methodDao;
   private final MethodVersionDao methodVersionDao;
+  private final RunSetDao runSetDao;
+
+  private static final List<String> SUPPORTED_SOURCES = new ArrayList<>(List.of("GitHub"));
```
Comment (cjllanwarne) [inline service/src/main/java/bio/terra/cbas/controllers/MethodsApiController.java:41]: Could this field be an `enum` in the openapi spec?

File: service/src/main/java/bio/terra/cbas/controllers/MethodsApiController.java [hunk 2]
```diff
@@ -50,6 +182,47 @@ public ResponseEntity<MethodListResponse> getMethods(
...
+    if (methodVersion == null || methodVersion.trim().isEmpty()) {
+      errors.add("method_version is required");
+    }
+
+    if (methodUrl == null || methodUrl.trim().isEmpty()) {
+      errors.add("method_url is required");
+    } else {
+      // verify that URL is valid, and it's host is supported
+      try {
+        URI uri = new URI(methodUrl);
```
Comment (cjllanwarne) [inline service/src/main/java/bio/terra/cbas/controllers/MethodsApiController.java:214]: I think you can do a lot of these required fields and format validations in `openapi` yaml. If you do that, our spring integration makes the values come in as URIs automatically without this manual string conversion

File: service/src/main/java/bio/terra/cbas/controllers/MethodsApiController.java [hunk 3]
```diff
@@ -50,6 +182,47 @@ public ResponseEntity<MethodListResponse> getMethods(
     return ResponseEntity.ok(new MethodListResponse().methods(methodDetails));
   }
 
+  public static List<String> validateMethod(PostMethodRequest methodRequest) {
```
Comment (salonishah11) [inline service/src/main/java/bio/terra/cbas/controllers/MethodsApiController.java:185]: TODO: Add checks for special characters

File: service/src/test/java/bio/terra/cbas/controllers/TestMethodsApiController.java [hunk 4]
```diff
@@ -191,6 +201,124 @@ void returnSpecificMethodVersion() throws Exception {
...
+
+    // verify that 'lastRunSetId' in method table was updated
+    ArgumentCaptor<RunSet> updateMethodCaptor = ArgumentCaptor.forClass(RunSet.class);
+    verify(methodDao).updateLastRunWithRunSet(updateMethodCaptor.capture());
+    assertEquals(postMethodResponse.getRunSetId(), updateMethodCaptor.getValue().runSetId());
+    assertEquals(
+        postMethodResponse.getMethodId(),
+        updateMethodCaptor.getValue().methodVersion().method().methodId());
+
+    // verify that 'lastRunSetId' in =method_version table was updated
```
Comment (salonishah11) [inline service/src/test/java/bio/terra/cbas/controllers/TestMethodsApiController.java:315]: TODO: Remove `=`

File: service/src/main/java/bio/terra/cbas/controllers/MethodsApiController.java (hunk 2 above)
Comment (salonishah11) [reply service/src/main/java/bio/terra/cbas/controllers/MethodsApiController.java:214]: Interesting. I had no idea. I will look into that

File: service/src/main/java/bio/terra/cbas/controllers/MethodsApiController.java (hunk 1 above)
Comment (salonishah11) [reply service/src/main/java/bio/terra/cbas/controllers/MethodsApiController.java:41]: Good point. I will make it an enum. So if it contains any other value the field will be `null` and then we can only check for that right? 🤔

File: service/src/main/java/bio/terra/cbas/controllers/MethodsApiController.java (hunk 2 above)
Comment (salonishah11) [reply service/src/main/java/bio/terra/cbas/controllers/MethodsApiController.java:214]: @cjllanwarne It seems that if properties are marked required in openapi.yaml and then not passed in the request body are received as null in the java code. So non-null checks might still be needed. For the `method_url` I tried adding `format: uri` but that doesn't check if the value is a URI or not. Am I missing something?

**Review by kpierre13 — APPROVED** [review]

File: service/src/main/java/bio/terra/cbas/controllers/MethodsApiController.java (hunk 3 above)
Comment (salonishah11) [reply service/src/main/java/bio/terra/cbas/controllers/MethodsApiController.java:188]: Done

**Review by cjllanwarne — APPROVED** [review]

**Pull request merged**